File: Java/Medium/TimeLimitExceeded/number-of-matching-subsequences_3.java
class Solution {
    public int numMatchingSubseq(String s, String[] words) {
        
        int count = 0;
        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }
        for (String word : map.keySet()) {
            if (isSubsequence(s, word)) {
                count += map.get(word);
            }
        }
        return count;
    }
    
    private boolean isSubsequence(String s, String word) {
        int i = 0;
        for (char c : s.toCharArray()) {
            if (i < word.length() && c == word.charAt(i)) {
                i++;
            }
        }
        return i == word.length();
    }
}

File: C++/Medium/Accepted/number-of-matching-subsequences_3.cpp
class Solution {
public:
    int numMatchingSubseq(string s, vector<string>& words) {
        
        int ans = 0;
        vector<vector<int>> pos(26);
        for (int i = 0; i < s.size(); i++) {
            pos[s[i] - 'a'].push_back(i);
        }
        for (auto& word: words) {
            int cur = -1;
            bool ok = true;
            for (auto& c: word) {
                auto& vec = pos[c - 'a'];
                auto it = lower_bound(vec.begin(), vec.end(), cur + 1);
                if (it == vec.end()) {
                    ok = false;
                    break;
                }
                cur = *it;
            }
            if (ok) ans++;
        }
        return ans;
    }
};